### ISSUE TITLE
Add support for third party concurrency libraries

### ISSUE THREAD

**Issue by RobertCraigie** [RobertCraigie/prisma-client-py#41]
## Problem

Not all python frameworks use the standard library `asyncio` module for running coroutines, we should add first class support for such frameworks

## Suggested solution

Not clear, don't have enough experience using other concurrency frameworks.

## Additional context

- [trio](https://github.com/python-trio/trio)
- [gevent](http://www.gevent.org/)
- [curio](https://github.com/dabeaz/curio)

**Comment by adriangb** [comment]
Probably just use [anyio](https://anyio.readthedocs.io). This gives you support for asyncio and trio w/ almost no code changes.

IMO it's not worth supporting gevent and curio.

**Comment by RobertCraigie** [comment]
I will look into that, thank you.

I don't have any experience using third part concurrency frameworks, why would it not be worth supporting gevent anc curio?

**Comment by adriangb** [comment]
> I will look into that, thank you.

Cool! I can probably help with the implementation if we end up aligning on the vision.

> I don't have any experience using third part concurrency frameworks, why would it not be worth supporting gevent and curio?

Firstly, from a technical perspective, supporting trio and asyncio via AnyIO is very easy. And you have to support asyncio anyway. There's nothing to "do" to support gevent: it monkey patches IO in the standard library. So users of gevent would just use the sync version and apply the gevent monkey patches. Finally, I think (could be wrong) that you can't / it's hard to support curio and asyncio at the same time.

On the practical side, the momentum in the Python ecosystem is towards asyncio + trio + whatever mix of them comes in the future (today that's anyio, tomorrow that may be changes to asyncio to support structured concurrency). I don't think anyone is using gevent or curio in new development, which is what async users of this library would probably be doing.

So TLDR, gevent should just work, curio is not worth supporting and asyncio+trio can easily be supported via AnyIO.

Again, if we aligning on the vision, I'm happy to make the (what I'm guessing is a simple PR) to add trio support via AnyIO.

**Comment by RobertCraigie** [comment]
Ah I see, thank you for clarifying that.

I am averse to adding dependencies and would rather work around issues if they are simple enough, however I will go against that if the work around is too complicated / increases maintainability burden too much.

I haven't used trio before so I don't know how much work would be required to support it, I will look into that before deciding whether or not to use AnyIO, thank you for your informative comment.

**Comment by adriangb** [comment]
> I am averse to adding dependencies and would rather work around issues if they are simple enough

Is this in reference to httpx or the concurrency backends?

**Comment by RobertCraigie** [comment]
@adriangb concurrency backends, I will happily add as many optional dependencies as required.

**Comment by RobertCraigie** [comment]
I hadn't actually considered making this feature optional, I will happily merge a PR that adds optional support for trio with AnyIO

**Comment by adriangb** [comment]
Makes sense, sorry my question was nonsensical, I got mixed up with our two threads 🤦‍♂️

**Comment by RobertCraigie** [comment]
@adriangb No worries!
